src/vs/base/browser/ui/dialog/dialog.ts

```diff
@@ -478,21 +478,33 @@ export class Dialog extends Disposable {
 
 	private rearrangeButtons(buttons: Array<string>, cancelId: number | undefined): ButtonMapEntry[] {
 		const buttonMap: ButtonMapEntry[] = [];
-		if (buttons.length === 0) {
+		if (buttons.length < 2) {
 			return buttonMap;
 		}
 
-		// Maps each button to its current label and old index so that when we move them around it's not a problem
-		buttons.forEach((button, index) => {
-			buttonMap.push({ label: button, index });
+		// Maps each button to its current label and old index
+		// so that when we move them around it's not a problem
+		buttons.forEach((label, index) => {
+			buttonMap.push({ label, index });
 		});
 
-		// macOS/linux: reverse button order if `cancelId` is defined
+		// macOS/linux: reverse button order and ensure cancel
+		// button is to the right of the primary button to match
+		// HIG per OS
```

Comment:
via https://github.com/microsoft/vscode/commit/19a5794679636e28c36e40bef33838a72cf86f15

I actually worry that moving the cancel button can break muscle memory. I tested on macOS Catalina where buttons are not arranged from top to bottom but in a row and found the cancel button next to the primary button (like on Linux) to be what native macOS apps do. 

I think the macOS HIGs maybe got updated (they must have when they added the top-bottom dialog stack for 3 buttons).

Reply:
This does look like a reasonable arrangement, *but* that may be partially because the delete is actually aligned to the far left, so it is more like you have the choice between cancel and save, and then there is a tertiary, fairly separate option to delete. I suspect that if they were just all in a row with no spacing, Cancel in the middle would feel off again. And when you actually have two positive, similarly weighted choices, plus a cancel, Like ["Commit", "Amend", "Cancel"], having this displayed as ["Amend", "Cancel", "Commit"] with no extra spacing is a bit weird. Do you know how the native APIs support the spacing? Or do they just always assume that you want your third button aligned to the opposite end of the dialog like the delete button here?

I still think having the explicitness of the API is worthwhile even without the actual order change, and I understand that changing something like this post fact is always controversial with users, so should be decided with care.

Reply:
We have no API in Electron to do above, so we have to stick with the row where buttons are closer. 

Again I think this is a nothing-burger, because most people by now will be on newer macOS where with 3 buttons, they stack from top to bottom.